Added support for slice.indices

ivirshup opened the pull request:
Fixes #1349 

(unless full support would now involve being able to pass `slice`s back to python?)

This PR adds support for the `slice.indices` method in compiled code. This was discussed [a bit on gitter](https://gitter.im/numba/numba?at=5da008c457c2517c6aea1084). It was my first time working with any of the extending code, so this could definitely use a looking over 🤞.

esc commented:
@ivirshup thanks for this, we appreciate you submitting this PR! I have now queued it for review.

esc commented on numba/tests/test_slices.py:
```diff
@@ -78,6 +83,25 @@ def test_slice_constructor(self):
+        """Test that a numba slice returns same result for .indices as a python one."""
+        N_TESTS = 500
+
+        lengths = np.random.randint(0, 100, N_TESTS)
```
I am not sure that random fuzzing is the correct approach. The problem with randomness is that it may be hard to repeat a failed test if the randomness manages to fuzz out an exotic and highly improbable combination of arguments. A well defined sequence of consistent input might be the wiser choice here.

ivirshup replied:
Agreed. I've replaced it with more comprehensive coverage on a smaller input space. This ends up being a larger set of test cases, and could probably be reduced a bit if runtime is an issue (it's ~0.25 seconds for `python -m numba.runtests -k "test_slice_indices"`). I've also added assertions to make sure the right exceptions were being thrown, and it turns out they weren't.

esc replied:
Maybe, the running time of our test-suite is an issue for sure. There is an open PR here that adds instrumentation to the test-suite:

https://github.com/numba/numba/pull/4640

I would tend to argue that 0.25 seconds is acceptable if it is a good balance. There are other test cases that are much worse offenders in terms of runtime.

esc replied:
> I've also added assertions to make sure the right exceptions were being thrown, and it turns out they weren't.

What we usually do, is have one set of known good inputs that we use to test the so-called 'happy-path'. Which is when the function does what it is supposed to. And another set of
tests that execute the function with input designed to raise exceptions. Usually, we check those using a `with self.asserttRaises` clause. Also, raising exceptions creates reference leaks, so you will need to call `self.disable_leak_check()` before checking those. There are numerous examples of this in the Numba test-code so it sholdn't be to hard to find something to crib off. For example this one:

https://github.com/numba/numba/blob/master/numba/tests/test_np_functions.py#L3067

ivirshup replied:
I could definitely add some explicit test cases here, and the `self.disable_leak_check()`.

From a fairly painful experience of trying to replicate slicing behaviour, I think it's hard to fully understand all possible cases. I'd like to keep the 'range of inputs against an oracle' test as it gives me much more confidence I haven't missed something.

ivirshup replied:
Do I need `disable_leak_check` here, as it doesn't currently subclass from `MemoryLeak`?

esc replied:
Yes, you need to subclass from `MemoryLeakMixin` to enable leak-checking.

ivirshup replied:
I've added the mixin and disabled checks on the tests which use exceptions.

esc commented:
it looks like CI is failing due to the type hints causing a `SyntaxError` on Python 2.7. It will be removing those to see if that fixes all the issues.

ivirshup commented:
@esc, I think I've addressed the issues you raised. Let me know what you think.

I also caught a few bugs related to exceptions not being thrown by the numba function. I think I've handled this right...

A slightly more general question, how can I benchmark this kind of function without including slice construction/ conversion time? I'm worried if I declare the slice in a numba function llvm might figure out it's a constant and optimize for it. Do I just have to measure conversion/ construction separately and subtract it?

ivirshup commented on numba/targets/slicing.py:
```diff
@@ -207,3 +207,30 @@ def slice_step_impl(context, builder, typ, value):
+        context.call_conv.return_user_exc(
+            builder, ValueError,
+            ("length should not be negative",)
+        )
```
Could I just wrap this in a `if utils.IS_PY3` block to deal with those python 2.7 test failures?

esc replied:
Not sure, why does this raise on 2.7 but not on the other python versions?

ivirshup replied:
This doesn't raise an error in 2.7. It does in python3. In python 2.7:

```python
>>> slice(None).indices(-10)
(-10, -10, 1)
```

From the "What's new" for python 3.4:

> slice.indices() no longer produces an OverflowError for huge values. As a consequence of this fix, slice.indices() now raises a ValueError if given a negative length; previously it returned nonsense values ([bpo-14794](https://bugs.python.org/issue14794)).

esc commented:
> A slightly more general question, how can I benchmark this kind of function without including slice construction/ conversion time? I'm worried if I declare the slice in a numba function llvm might figure out it's a constant and optimize for it. Do I just have to measure conversion/ construction separately and subtract it?

@stuartarchibald probably has an answer for this.

esc commented:
@ivirshup thanks for addressing the review remarks. I have left a few more comments for your consideration. The PR is definitely going in the right direction, good job!

ivirshup commented:
@esc, I'm having some difficulty with this python2 stuff. I'll lay out where I'm at, and then what I'm thinking.

In python2, `slice(...).indices({negative_integer})` doesn't throw an error. It does, (based on the python change log) return nonsense results, since a negative length isn't really a valid case. In the spirit of compatibility I've tried to replicate those nonsense results, but there's a case where I'm not sure I can replicate without a lot of effort. When `slice.stop is None`, `slice.step < 0`, and `length < 0`, python2 returns `maxint + length` for the `stop` value from `slice.indices`. Here's an example:

```python
>>> slice(None, None, -2).indices(-1)
(-2, 9223372036854775807, -2)
```

Since numba slices don't store `None`,  just integer values which give equivalent slicing results, I can't think of how to tell when this case occurs without modifying the constructor and kinds of values numba slices hold.

To me, this seems like probably too much work to replicate what's pretty seems like a bug (or at least undefined behaviour) in python2, especially since python2 won't even be supported in a few months. As an alternative, I'm thinking I'll just not test the equivalency of results for `slice.indices` in python2 when length is negative, and keeping it a `ValueError` (like in python3) instead of returning nonsense.

Thoughts?

stuartarchibald commented:
@ivirshup Given Python 2 has about 2-3 months of support left with Numba I'm inclined to not worry about it too much either. I don't think it's worth trying to accurately reproduce this result as I doubt code is going to be relying on it. As suggested, perhaps just make this work on Python 3 correctly and then skip the unusual result case in the unit tests on Python 2 and leave a note about why? Thanks.

ivirshup commented:
@stuartarchibald, @esc, thanks for the feedback! I've updated the comment in the test to give more background on why those cases are disabled.

Is there anything else I should change here?

esc commented:
@ivirshup thanks for the inquiry, I'll queue it for the next round of inspection.

esc commented on numba/tests/test_slices.py:
```diff
@@ -78,6 +82,62 @@ def test_slice_constructor(self):
+
+        cfunc = jit(nopython=True)(slice_indices)
+
+        for s, l in product(slices, lens):
```
This loop is a bit tricky to read and understand a mixes several concerns. It would be good to simplify this. But I am not yet sure how to make this more legible.  How about the following:

```
            if l < 0 and not utils.IS_PY3:
                # Passing a negative length to slice.indices in python2 is
                # undefined. See https://bugs.python.org/issue14794#msg174678
                continue
            try:
                expected = slice_indices(s, l)
            except Exception as e:
                with self.assertRaises(type(e)):
                    cfunc(s, l)
            else:
                    self.assertPreciseEqual(expected, cfunc(s, l))
```
That way it is a bit easier to read the flow:
* First, skip the length less than zero on Python 2 entirely
* Then try to compare Numba and CPython, if it's an exception, make sure the same on is raised and otherwise check the result.

Makes sense?

ivirshup replied:
Yup, your way is more legible. I've switched to this.

ivirshup commented on numba/tests/test_slices.py:
```diff
@@ -78,6 +82,62 @@ def test_slice_constructor(self):
+                chain(range(-5, 5), (None,))
+            )
+        )
+        lens = range(-2, 3)
```
👍, fixed

esc commented:
I tried to run the code with the `disable_leak_check` commented out and the tests don't seem to be leaking anything. @stuartarchibald is this to be expected?

esc commented:
@ivirshup o.k. so, since the function you are testing doesn't allocate anything that could leak, raising the exceptions during testing doesn't leak either. Sorry for the confusion. My  recommendation would be to remove the `disable_leak_check` but leave in the `MemoryLeakMixin` to protect agains any regressions.

ivirshup commented:
@esc, no problem. I've pushed those changes.

esc commented:
@ivirshup thanks for your persistence, I think this is now almost ready for merge. I'd like to request a final review from @stuartarchibald or @sklam and it is good to go.

stuartarchibald commented on numba/typing/builtins.py:
```diff
@@ -726,6 +726,11 @@ def resolve_stop(self, ty):
+    @bound_function("slice.indices")
+    def resolve_indices(self, ty, args, kws):
+        assert not kws
+        return signature(types.UniTuple(types.intp, 3), types.intp)
```
This needs a guard to check that `args[0]` is a `types.Integer` and only 1 arg is present e.g.:
```python
        val = args[0]
        if len(args==1) and isinstance(val, types.Integer):
            return signature(types.UniTuple(types.intp, 3), types.intp)
```
without this there's an implicit guarantee that this function will consume an `intp` and cast anything to it and return a 3-tuple. i.e. this "works", when it really should not.

```python
from numba import njit

a  = slice(1, 9, 2)

@njit
def foo(x):
    return x.indices(12.34)

print(foo(a))
```
as does this:
```python
from numba import njit

a  = slice(1, 9, 2)

@njit
def foo(x):
    return x.indices(10, 20, 30, 40, 50, 60)

print(foo(a))
```
both of the above perhaps ought to become part of the unit tests?

ivirshup replied:
Ah, now I see what's up with the convention of `[ty] = args`.

Should these throw typing errors, or should I try to match the python errors?

stuartarchibald replied:
I'm pretty sure they won't need to explicitly throw anything, if a typing template doesn't return a signature it means the type combination being compiled isn't supported (and an automatically generated TypingError should appear). Perhaps check that is the case first and if the error message makes sense, if so, it's just a question of adding tests to make sure that that happens correctly. If the messages are weird it's fine to instead have an `else` branch raise a TypingError with a specific message that you feel is more suitable.

ivirshup replied:
Based on a few other `bound_function` definitions ([example](https://github.com/numba/numba/blob/8c6658375c1f8fe50e1a5ccd11d4e7bf5a8053de/numba/typing/arraydecl.py#L299)) I tried raising the error python would have thrown. This gets overwritten with a typing error, but at least the message is included. Here's an reduced example of what this looks like:

```python
>>> from numba import njit
>>> slice_indices = lambda s, *indargs: s.indices(*indargs)
>>> cslice_indices = njit(slice_indices)

>>> slice_indices(slice(None), 1.2)
TypeError                                 Traceback (most recent call last)
...
TypeError: 'float' object cannot be interpreted as an integer

>>> cslice_indices(slice(None), 1.2)
TypingError                               Traceback (most recent call last)
...
TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Internal error at <numba.typeinfer.CallConstraint object at 0x625301e50>.
'float64' object cannot be interpreted as an integer
...
```

Thoughts?

stuartarchibald replied:
Seems good, thanks.

ivirshup commented on numba/tests/test_slices.py:
```diff
@@ -78,6 +82,61 @@ def test_slice_constructor(self):
+        with self.assertRaises(ValueError):
+            cslice_indices(slice(None, None, 0), 0)
+
+        for args, length, expected in [
```
I interpreted @esc's [comment above](https://github.com/numba/numba/pull/4707#discussion_r335344896) as asking for explicit test cases. Reading it again, I might have been wrong about that. I don't think this covers anything else, but is more for documentation of what should throw an error.

stuartarchibald replied:
ok, I'm not sure these are needed, they don't seem to add anything more than what is already present.

ivirshup replied:
Removed them

ivirshup commented on numba/targets/slicing.py:
```diff
@@ -207,3 +207,30 @@ def slice_step_impl(context, builder, typ, value):
         return context.get_constant(types.intp, 1)
+
+
+@lower_builtin("slice.indices", types.SliceType, types.Integer)
```
Huh. I really hope not.

ivirshup commented on numba/tests/test_slices.py:
```diff
@@ -78,6 +82,61 @@ def test_slice_constructor(self):
+
+        # slice.indices shouldn't allow negative lengths
+        if utils.IS_PY3:
+            with self.assertRaises(ValueError):
```
Updated this to check the contents of the numba exception against the contents of the python one, instead of using a literal string. That sound good?

ivirshup commented:
Thanks for the feedback!
